In [7]:
import pandas as pd


In [8]:
df = pd.read_csv('/Users/hatice/Documents/proje/online_retail_II.csv', encoding='latin1')  # dosyada özel karakterler var, utf-8 hata veriyor


In [9]:
df.head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/10 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/10 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/10 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/10 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/10 8:26,3.39,17850.0,United Kingdom
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,12/1/10 8:26,7.65,17850.0,United Kingdom
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,12/1/10 8:26,4.25,17850.0,United Kingdom
7,536366,22633,HAND WARMER UNION JACK,6,12/1/10 8:28,1.85,17850.0,United Kingdom
8,536366,22632,HAND WARMER RED POLKA DOT,6,12/1/10 8:28,1.85,17850.0,United Kingdom
9,536368,22960,JAM MAKING SET WITH JARS,6,12/1/10 8:34,4.25,13047.0,United Kingdom


In [10]:
df['InvoiceDate'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 541910 entries, 0 to 541909
Series name: InvoiceDate
Non-Null Count   Dtype 
--------------   ----- 
541910 non-null  object
dtypes: object(1)
memory usage: 4.1+ MB


In [11]:
df.info()
#InvoiceDate dtype object 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Invoice      541910 non-null  object 
 1   StockCode    541910 non-null  object 
 2   Description  540456 non-null  object 
 3   Quantity     541910 non-null  int64  
 4   InvoiceDate  541910 non-null  object 
 5   Price        541910 non-null  float64
 6   Customer ID  406830 non-null  float64
 7   Country      541910 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [12]:
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'], format='%m/%d/%y %H:%M')

#InvoiceDate kolonundaki text halinde duran tarihleri,Python'un gerçek bir tarih/saat nesnesi olarak tanıyabileceği bir formata çeviriyoruz.

In [13]:
df['InvoiceDate'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 541910 entries, 0 to 541909
Series name: InvoiceDate
Non-Null Count   Dtype         
--------------   -----         
541910 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 4.1 MB


In [14]:
df.info()
#InvoiceDate dtype datetime oldu

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      541910 non-null  object        
 1   StockCode    541910 non-null  object        
 2   Description  540456 non-null  object        
 3   Quantity     541910 non-null  int64         
 4   InvoiceDate  541910 non-null  datetime64[ns]
 5   Price        541910 non-null  float64       
 6   Customer ID  406830 non-null  float64       
 7   Country      541910 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [15]:
# verisetinde boş satır var mı?  
print(df.isnull().sum())

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64


In [16]:
#Description      1454
#Customer ID    135080
# Customer ID   olmayan satırlar var dolayısıyla burada müşteri bazlı bir analiz yapmamız konusunda problem olabilir

In [17]:
df[df['Customer ID'].isnull()].head(10)
#Customer ID olmayan veya cutomer ID NaN olan ilk 10 satırı getirdi

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,NaN,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,NaN,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,NaN,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,NaN,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,NaN,United Kingdom


In [18]:
# Customer ID'si olmayan satirlari cikaralım . Onlar müşüteri bazlı yaptığımız analizde sorun olabilir.
df = df.dropna(subset=['Customer ID']).copy()

df.dropna(subset=['Customer ID']) → "Customer ID kolonu NaN olan satırları tamamen çıkar demiş oluyor.

subset=['Customer ID'] ise "sadece bu kolona bak, sadece bu kolonu boş olan satırları at" komutunu ifade ediyor.
Eğer subset belirtmeseydik, pandas herhangi bir kolonu boş olan satırı atardı (ki bizim Description'da da 1454 boş vardı, onları atmak istemeyiz).
.copy() dememizin sebebi ise bir nevi işimizi garantiye almak : pandas'ta bazen bir filtrelemeden sonra elde ettiğin tablo, orijinal tablonun view hali olur yani gerçek bağımsız bir kopyası olmaz. Bu durumda ileride o tabloyu değiştirmeye çalışırsan pandas sana SettingWithCopyWarning diye bir uyarı verebilir dolayısıyla hataya yol açabilir. .copy() diyerek "bu artık tamamen bağımsız, yeni bir tablo olsun" diyerek garanti altına alıyoruz.

In [19]:
df['Customer ID'] = df['Customer ID'].astype(int)  # Customer ID artık integer

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 406830 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      406830 non-null  object        
 1   StockCode    406830 non-null  object        
 2   Description  406830 non-null  object        
 3   Quantity     406830 non-null  int64         
 4   InvoiceDate  406830 non-null  datetime64[ns]
 5   Price        406830 non-null  float64       
 6   Customer ID  406830 non-null  int64         
 7   Country      406830 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 27.9+ MB


In [21]:
# Datasette şöyle bir ayrım var. Invoice sütunundaki değerler eğer C ile başlıyorsa bu iptal edilen sipariş anlamına geliyor.
df['is_cancellation'] = df['Invoice'].astype(str).str.startswith('C')  # Iptal/iade isareti: Invoice "C" ile basliyorsa 


In [22]:
df.info()   # is_cancellation artık yeni bir kolon olarak eklendi. 

<class 'pandas.core.frame.DataFrame'>
Index: 406830 entries, 0 to 541909
Data columns (total 9 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   Invoice          406830 non-null  object        
 1   StockCode        406830 non-null  object        
 2   Description      406830 non-null  object        
 3   Quantity         406830 non-null  int64         
 4   InvoiceDate      406830 non-null  datetime64[ns]
 5   Price            406830 non-null  float64       
 6   Customer ID      406830 non-null  int64         
 7   Country          406830 non-null  object        
 8   is_cancellation  406830 non-null  bool          
dtypes: bool(1), datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 28.3+ MB


In [23]:
df['is_cancellation'].sum()     

8905

In [24]:
df['Invoice'].astype(str).str.startswith('C').sum()

8905

In [25]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,is_cancellation
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,False
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,False
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,False
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,False
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,False


In [26]:
# Satir tutari diye yeni bir kolon ekliyoruz. Çünkü bu ilgili satırdaki siparişin toplam tutarını belirtecek. 
df['line_total'] = df['Quantity'] * df['Price']

In [27]:
df.head()

#Örneğin 6*2.55= 15.3 gibi bir sonuç çıkıyordu. Bunu bir kolon haline getirdik ve o satırın toplam tutarı artık yazılı.

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,is_cancellation,line_total
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,False,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,False,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,False,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,False,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,False,20.34


veriseti toplam kaç günlük ya da aylık bir zamanı kapsıyo bunu öğrenelim

In [28]:

df['InvoiceDate'].max()


Timestamp('2011-12-09 12:50:00')

In [29]:
df['InvoiceDate'].min()

Timestamp('2010-12-01 08:26:00')

In [30]:
print("Toplam gün sayısı:", (df['InvoiceDate'].max() - df['InvoiceDate'].min()).days)

Toplam gün sayısı: 373


In [31]:
ay_sayisi = 373 / 30.44 
print(round(ay_sayisi, 1))

#Yaklaşık 12 aylık bir zamanı kapsıyor bu veriseti 

12.3


Veri seti yaklaşık 12.3 ay kapsıyor (2010-12 → 2011-12). Son ~2.5 ayı "gelecek" olarak ayırdık, geri kalan ~10 ayı "geçmiş" (feature çıkarma) için kullandık. Bu oran keyfi değil — yeterince uzun bir gelecek pencere olmalı ki müşterilerin iptal davranışını gözlemleyebilelim, ama feature çıkaracak geçmiş de yeterince uzun olmalı (aksi halde müşteri davranışını iyi temsil edemez).

Şimdi yavaş yavaş modelleme süreçlerine geçiyoruz. Bu yüzden verisetini iki zaman çizgisine bölüyoruz. 
bir 10 aylık bir kısım 
diğeri 2 aylık bir kısım 

In [32]:
SPLIT_DATE = pd.Timestamp('2011-10-01')



In [33]:
past = df[df['InvoiceDate'] < SPLIT_DATE].copy()
future = df[df['InvoiceDate'] >= SPLIT_DATE].copy()

groupby('Customer ID') her müşteri için satırları gruplayacak,
.agg() her gruba farklı bir hesaplama uygulayacak.
Örneğin total_orders=('Invoice', 'nunique') demek: "her müşterinin kaç farklı fatura numarası var, onu say" demek oluyor.

In [34]:
features = past.groupby('Customer ID').agg(
    total_orders=('Invoice', 'nunique'),        # kac farkli siparis verdi
    total_line_items=('Invoice', 'count'),      # toplam kac urun satiri
    total_spend=('line_total', 'sum'),          # toplam harcama
    avg_order_value=('line_total', 'mean'),     # ortalama satir tutari
    distinct_products=('StockCode', 'nunique'), # kac farkli urun almis
    first_purchase=('InvoiceDate', 'min'),
    last_purchase=('InvoiceDate', 'max'),
).reset_index()

print(features)


      Customer ID  total_orders  total_line_items  total_spend  \
0           12346             2                 2         0.00   
1           12347             5               124      2790.86   
2           12348             4                31      1797.24   
3           12350             1                17       334.40   
4           12352            10                80      1233.68   
...           ...           ...               ...          ...   
3653        18280             1                10       180.60   
3654        18281             1                 7        80.82   
3655        18282             2                 8        98.76   
3656        18283            10               447      1120.67   
3657        18287             1                29       765.28   

      avg_order_value  distinct_products      first_purchase  \
0            0.000000                  1 2011-01-18 10:01:00   
1           22.506935                 82 2010-12-07 14:57:00   
2           57.

is_cancellation True/False (yani 1/0) olduğu için .sum() toplam iptal sayısını verir. Oranı (cancel_count / total_txn) hesaplayıp ayrı bir feature olarak ekledik — "bu müşterinin geçmişte iptal etme alışkanlığı ne kadar yoğun" sorusuna cevap.  

total_txn 'i transaction oalrak düşünebiliriz. Yani transaction işlem anlamına geliyor dolayısıyla burada ilgili müşterinin yaptığıı total işlem sayısını gözlemliyoruz. 

Diyelim bir müşterinin geçmiş dönemde toplam 50 satırı var (bunların 45'i normal alışveriş, 5'i iptal/iade):

total_txn = 50 (toplam işlem sayısı)
cancel_count = 5 (bunlardan kaçı iptaldi)


In [35]:
cancel_counts = past.groupby('Customer ID')['is_cancellation'].agg(
    cancel_count='sum', total_txn='count'
).reset_index()

print(cancel_counts)

      Customer ID  cancel_count  total_txn
0           12346             1          2
1           12347             0        124
2           12348             0         31
3           12350             0         17
4           12352            10         80
...           ...           ...        ...
3653        18280             0         10
3654        18281             0          7
3655        18282             1          8
3656        18283             0        447
3657        18287             0         29

[3658 rows x 3 columns]


In [36]:
cancel_counts['past_cancel_rate'] = cancel_counts['cancel_count'] / cancel_counts['total_txn']

print(cancel_counts['past_cancel_rate'])

0       0.500
1       0.000
2       0.000
3       0.000
4       0.125
        ...  
3653    0.000
3654    0.000
3655    0.125
3656    0.000
3657    0.000
Name: past_cancel_rate, Length: 3658, dtype: float64


In [37]:
features = features.merge(cancel_counts[['Customer ID', 'cancel_count', 'past_cancel_rate']], on='Customer ID', how='left')

print(features)

# on='Customer ID' demek Customer ID'ye göre eşleştir demek oluyor. Bir nevi Customer ID key .
# how='left' ise bir nevi birleşme türü olarak kullanıldı. eşleşme bulunamadığında ne olacağını ve hangi tablonun esas alınacağını belirliyor.

      Customer ID  total_orders  total_line_items  total_spend  \
0           12346             2                 2         0.00   
1           12347             5               124      2790.86   
2           12348             4                31      1797.24   
3           12350             1                17       334.40   
4           12352            10                80      1233.68   
...           ...           ...               ...          ...   
3653        18280             1                10       180.60   
3654        18281             1                 7        80.82   
3655        18282             2                 8        98.76   
3656        18283            10               447      1120.67   
3657        18287             1                29       765.28   

      avg_order_value  distinct_products      first_purchase  \
0            0.000000                  1 2011-01-18 10:01:00   
1           22.506935                 82 2010-12-07 14:57:00   
2           57.

In [38]:
features.head()


,Customer ID,total_orders,total_line_items,total_spend,avg_order_value,distinct_products,first_purchase,last_purchase,cancel_count,past_cancel_rate
0,12346,2,2,0.00,0.000000,1,2011-01-18 10:01:00,2011-01-18 10:17:00,1,0.500
1,12347,5,124,2790.86,22.506935,82,2010-12-07 14:57:00,2011-08-02 08:48:00,0,0.000
2,12348,4,31,1797.24,57.975484,22,2010-12-16 19:09:00,2011-09-25 13:13:00,0,0.000
3,12350,1,17,334.40,19.670588,17,2011-02-02 16:01:00,2011-02-02 16:01:00,0,0.000
4,12352,10,80,1233.68,15.421000,47,2011-02-16 12:33:00,2011-09-28 14:58:00,10,0.125


recency_days = "referans tarihimize (1 Ekim) göre son alışverişin üzerinden kaç gün geçmiş" — düşük değer = yakın zamanda aktif müşteri. 

tenure_days = "müşteri ne kadar süredir bizde alışveriş yapıyor" — bu  RFM (Recency-Frequency-Monetary) analizinin mantığı, 

In [39]:
REFERENCE_DATE = SPLIT_DATE
#son_gün_sayısı
features['recency_days'] = (REFERENCE_DATE - features['last_purchase']).dt.days 
#gün bazlı bağlılık süresi . Müşteri ne kadar süredir bizden alışveriş yapıyor
features['tenure_days'] = (REFERENCE_DATE - features['first_purchase']).dt.days
#bir müşterinin sipariş başına ortalama kaç ürün satırı aldığını gösteriyor 
#yani "bu müşteri genelde tek seferde az mı çok mu ürün alıyor"
features['avg_items_per_order'] = features['total_line_items'] / features['total_orders']

x.mode().iloc[0] = "bu müşterinin en sık göründüğü ülke" (bir müşteri neredeyse hep aynı ülkeden alışveriş yapar ama garantiye almak için mod/en sık değeri alalım).

In [40]:
top_country = past.groupby('Customer ID')['Country'].agg(lambda x: x.mode().iloc[0]).reset_index()
top_country.columns = ['Customer ID', 'top_country']
features = features.merge(top_country, on='Customer ID', how='left')

In [41]:
print(SPLIT_DATE)

2011-10-01 00:00:00


In [42]:
print(features['avg_items_per_order'])

0        1.00
1       24.80
2        7.75
3       17.00
4        8.00
        ...  
3653    10.00
3654     7.00
3655     4.00
3656    44.70
3657    29.00
Name: avg_items_per_order, Length: 3658, dtype: float64


In [43]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,is_cancellation,line_total
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,False,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,False,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,False,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,False,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,False,20.34


In [44]:
print(features)

      Customer ID  total_orders  total_line_items  total_spend  \
0           12346             2                 2         0.00   
1           12347             5               124      2790.86   
2           12348             4                31      1797.24   
3           12350             1                17       334.40   
4           12352            10                80      1233.68   
...           ...           ...               ...          ...   
3653        18280             1                10       180.60   
3654        18281             1                 7        80.82   
3655        18282             2                 8        98.76   
3656        18283            10               447      1120.67   
3657        18287             1                29       765.28   

      avg_order_value  distinct_products      first_purchase  \
0            0.000000                  1 2011-01-18 10:01:00   
1           22.506935                 82 2010-12-07 14:57:00   
2           57.

In [45]:
print(SPLIT_DATE)

2011-10-01 00:00:00


Hangi müşteri gelecekte iptal etti.

sadece gelecek donemde en az 1 islemi olan musterileri dahil ediyoruz (yoksa etiket "iptal edecek mi" degil, gizlice "aktif olacak mi" sorusunu da olcer)

In [46]:
future_active_customers = set(future['Customer ID'].unique())

In [47]:
print(future_active_customers)

{16385, 16386, 16389, 16393, 16394, 16395, 16399, 16401, 16406, 16407, 16409, 16411, 16412, 16413, 16414, 16415, 16416, 16418, 16422, 16423, 16426, 16430, 16432, 16433, 16434, 16438, 16440, 16441, 16442, 16444, 16445, 16446, 16454, 16455, 16458, 16460, 16462, 16463, 16464, 16466, 16469, 16470, 16474, 16477, 16478, 16479, 16482, 16488, 16491, 16495, 16496, 16497, 16500, 16504, 16505, 16506, 16511, 16513, 16515, 16516, 16518, 16519, 16523, 16525, 16528, 16531, 16532, 16533, 16535, 16539, 16542, 16549, 16554, 16555, 16556, 16557, 16558, 16560, 16561, 16563, 16566, 16569, 16570, 16571, 16591, 16592, 16593, 16595, 16596, 16597, 16600, 16601, 16602, 16607, 16609, 16612, 16616, 16618, 16620, 16621, 16626, 16628, 16633, 16637, 16638, 16639, 16642, 16648, 16652, 16653, 16655, 16656, 16657, 16660, 16666, 16667, 16668, 16669, 16670, 16671, 16672, 16674, 16676, 16678, 16680, 16682, 16684, 16685, 16686, 16688, 16690, 16693, 16700, 16701, 16705, 16706, 16708, 16710, 16711, 16712, 16713, 16717, 16719

In [48]:
features = features[features['Customer ID'].isin(future_active_customers)].copy()


In [49]:
future_cancel_customers = set(future[future['is_cancellation']]['Customer ID'].unique())

In [50]:
features['label_will_cancel'] = features['Customer ID'].isin(future_cancel_customers).astype(int)


In [51]:
print(features['label_will_cancel'].value_counts())

label_will_cancel
0    1350
1     537
Name: count, dtype: int64


In [52]:
# Modele girmeyecek kolonlari kimlik, ara hesaplama icin kullanilan tarih kolonlarini çıkaralım
model_df = features.drop(columns=['Customer ID', 'cancel_count'])

In [53]:
print(df.columns)

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country', 'is_cancellation', 'line_total'],
      dtype='object')


In [54]:
# Kategorik degiskeni (top_country) sayisal hale getir
model_df = pd.get_dummies(model_df, columns=['top_country'], drop_first=True)

In [55]:
print(model_df.shape)


(1887, 35)


In [56]:
model_df.head()

,total_orders,total_line_items,total_spend,avg_order_value,distinct_products,first_purchase,last_purchase,past_cancel_rate,recency_days,tenure_days,...,top_country_Malta,top_country_Netherlands,top_country_Norway,top_country_Poland,top_country_Portugal,top_country_Singapore,top_country_Spain,top_country_Sweden,top_country_Switzerland,top_country_United Kingdom
1,5,124,2790.86,22.506935,82,2010-12-07 14:57:00,2011-08-02 08:48:00,0.000000,59,297,...,False,False,False,False,False,False,False,False,False,False
4,10,80,1233.68,15.421000,47,2011-02-16 12:33:00,2011-09-28 14:58:00,0.125000,2,226,...,False,False,True,False,False,False,False,False,False,False
8,2,57,2753.08,48.299649,53,2011-01-18 09:50:00,2011-04-08 12:33:00,0.000000,175,255,...,False,False,False,False,True,False,False,False,False,False
9,1,12,484.86,40.405000,12,2011-07-12 10:04:00,2011-07-12 10:04:00,0.000000,80,80,...,False,False,False,False,False,False,False,False,False,False
10,4,145,3459.93,23.861586,135,2011-01-12 12:43:00,2011-06-03 12:26:00,0.013793,119,261,...,False,False,False,False,False,False,False,False,False,False


In [57]:
model_df.info()

# top_country bu kadar fazla görünüyor çünkü One-Hot Encoding Mantığı
#get_dummies ile top_country kolonundaki her bir farklı ülke değeri için ayrı bir 0/1 (True/False) kolonu oluşturduk
# Yani orijinal tabloda tek bir kolon vardı (top_country, içinde "Germany", "France", "Italy" gibi metin değerleri), 
#şimdi bu tek kolon, her ülke için ayrı bir sütuna dönüştü

#Makine öğrenmesi modelleri (Logistic Regression, Random Forest vb.) sadece sayılarla çalışır — "Germany" veya "Italy" gibi 
#metin değerlerini doğrudan anlayamaz, matematiksel işlem yapamaz. Bu yüzden kategorik (metin) bir değişkeni, 
#modelin anlayabileceği 0/1'lerden oluşan bir forma çevirmemiz gerekiyor — buna one-hot encoding deniyor.

<class 'pandas.core.frame.DataFrame'>
Index: 1887 entries, 1 to 3657
Data columns (total 35 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   total_orders                 1887 non-null   int64         
 1   total_line_items             1887 non-null   int64         
 2   total_spend                  1887 non-null   float64       
 3   avg_order_value              1887 non-null   float64       
 4   distinct_products            1887 non-null   int64         
 5   first_purchase               1887 non-null   datetime64[ns]
 6   last_purchase                1887 non-null   datetime64[ns]
 7   past_cancel_rate             1887 non-null   float64       
 8   recency_days                 1887 non-null   int64         
 9   tenure_days                  1887 non-null   int64         
 10  avg_items_per_order          1887 non-null   float64       
 11  label_will_cancel            1887 non-null   int

In [58]:
print(model_df.columns.tolist())

['total_orders', 'total_line_items', 'total_spend', 'avg_order_value', 'distinct_products', 'first_purchase', 'last_purchase', 'past_cancel_rate', 'recency_days', 'tenure_days', 'avg_items_per_order', 'label_will_cancel', 'top_country_Austria', 'top_country_Belgium', 'top_country_Channel Islands', 'top_country_Cyprus', 'top_country_Czech Republic', 'top_country_Denmark', 'top_country_EIRE', 'top_country_Finland', 'top_country_France', 'top_country_Germany', 'top_country_Iceland', 'top_country_Italy', 'top_country_Japan', 'top_country_Malta', 'top_country_Netherlands', 'top_country_Norway', 'top_country_Poland', 'top_country_Portugal', 'top_country_Singapore', 'top_country_Spain', 'top_country_Sweden', 'top_country_Switzerland', 'top_country_United Kingdom']


In [59]:
features['top_country'].value_counts()

top_country
United Kingdom     1707
Germany              46
France               43
Belgium              15
Spain                12
Portugal              8
Norway                6
Switzerland           6
Italy                 6
Netherlands           5
Sweden                4
Channel Islands       4
Finland               4
Denmark               3
Australia             3
Austria               3
EIRE                  2
Cyprus                2
Poland                2
Japan                 2
Iceland               1
Czech Republic        1
Singapore             1
Malta                 1
Name: count, dtype: int64

Ülke Dağılımı Aşırı Dengesiz 
1887 müşterinin 1707'si (yaklaşık %90'ı) United Kingdom. Geri kalan 23 ülke, toplamda sadece 180 müşteriye dağılmış — çoğu ülkede tek haneli sayıda müşteri var (Malta: 1, Japonya: 2, İzlanda: 1 gibi).

Bunun 23 ayrı kolon olarak modele girmesi neden sorunlu:

Malta'dan 1 müşteri var — bu tek örnekten model hiçbir genel örüntü öğrenemez, sadece o tek satırın gürültüsünü ezberler (overfitting).

1887 satırlık bir veriye 23 tane "neredeyse hep False" kolon eklemek, modelin gerçekten önemli sinyalleri (past_cancel_rate, recency_days gibi) bulmasını zorlaştırabilir.

Yukarıdaki satırlarda gördüğümüz üzere get_dummies ile herbir ülkeyi farklı kolonlara bölmüştük 

bunun yerine ne yapmalı: 23 ülke kolonu yerine, tek bir basit kolon kullanalım: "UK mi, değil mi" olarak . Çünkü 1887 müşterinin 1707'si (yaklaşık %90'ı) United Kingdom 

In [60]:
model_df = features.drop(columns=[
    'Customer ID', 'cancel_count', 'top_country', 
    'total_line_items', 'first_purchase', 'last_purchase'
])
model_df['is_uk'] = (features['top_country'] == 'United Kingdom').astype(int)

print(model_df.shape)
print(model_df.dtypes)

(1887, 10)
total_orders             int64
total_spend            float64
avg_order_value        float64
distinct_products        int64
past_cancel_rate       float64
recency_days             int64
tenure_days              int64
avg_items_per_order    float64
label_will_cancel        int64
is_uk                    int64
dtype: object


Girdi (X) ve Hedef (y)

model_df tablosundaki tüm kolonları ikiye ayırıyoruz:

X → modelin görüp öğreneceği özellikler (total_orders, total_spend, past_cancel_rate vb.) — label_will_cancel hariç hepsi

y → modelin tahmin etmeye çalışacağı hedef değişken (0 veya 1, iptal edip etmeyeceği)

In [61]:
X = model_df.drop(columns=['label_will_cancel'])
y = model_df['label_will_cancel']

# Veriyi %80 eğitim, %20 test olarak ikiye bölüyoruz — modelin hiç görmediği verilerle test edilmesini sağlamak için
# stratify=y, her iki parçada da iptal oranının  korunmasını garanti ediyor.


from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


# Random Forest Modeli
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=200, class_weight='balanced', random_state=42, max_depth=6)
rf.fit(X_train, y_train)  # olceklenmemis veriyle

#Test Setinde Tahmin ve Değerlendirme
from sklearn.metrics import classification_report
y_pred_rf = rf.predict(X_test)
print(classification_report(y_test, y_pred_rf))

              precision    recall  f1-score   support

           0       0.80      0.79      0.80       270
           1       0.50      0.52      0.51       108

    accuracy                           0.71       378
   macro avg       0.65      0.65      0.65       378
weighted avg       0.72      0.71      0.71       378



Veriyi girdi/hedef olarak ayırdık → eğitim/test olarak böldük → (kullanılmayan bir ölçekleme adımı var) → Random Forest modelini eğitim verisiyle eğittik → test verisinde tahmin yaptırdık → tahminlerin ne kadar doğru olduğunu bir raporla ölçtük.

model çıktısını LLM'e hazırlamaya çalışalım. Çünkü Modelin Ürettiği Şey, Tek Başına Kullanılabilir Değil  
Random Forest modelimiz, her müşteri için sadece şunu üretiyor:

Customer ID: 16558, risk_probability: 0.86

Bu, bir veri bilimci veya teknik biri için anlamlı bir çıktı  ama kampanya yöneticisi veya operasyon ekibi için anlamlı bir çıktı değil.

"%86 risk" gördüğünde, neden bu kadar yüksek olduğunu bilmiyor (hangi davranış bunu tetikledi?)

Bu bilgiyle ne yapması gerektiğini bilmiyor (bir e-posta mı göndersin, indirim mi versin, hiçbir şey yapmasın mı?)

Yani model bize "kim riskli" sorusuna cevap veriyor, ama "neden riskli" ve "ne yapmalıyım" sorularına cevap vermiyor


Burada projeye bir LLM katmanı ekliyoruz . LLM Katmanının Rolü: Sayıyı Aksiyona Çevirmek 

LLM'e model çıktısını (risk skoru + davranışsal özellikler) verdiğimizde, LLM bunu:

Yorumluyor (hangi özellikler bu riski işaret ediyor — "yüksek ürün çeşitliliği, deneme-iade davranışına işaret ediyor" gibi)

Doğal dile çeviriyor (operasyon ekibinin anlayacağı, teknik jargon içermeyen bir dille)
Somut bir aksiyon öneriyor ("bu müşteriye şu tür bir e-posta gönderilsin" gibi)

Bu aşamada, yukarıdaki prompt'ları programatik olarak Anthropic API'sine göndermek yerine, bilinçli olarak manuel bir yol seçtim: prompt'ları kod ile otomatik olarak müşteri verisiyle doldurdum, ama LLM'e gönderme ve yanıt alma adımını Claude.ai arayüzü üzerinden elle yaptım.

Bunun sebebi teknik bir sınırlama değil, bir kaynak kullanım kararıydı — prototip/geliştirme aşamasında gereksiz API maliyeti oluşturmadan, prompt tasarımını doğrulamak istedim. Üretim ortamına taşınacak bir sistemde, bu son adım (`client.messages.create(...)` çağrısı) birkaç satır kodla otomatikleştirilebilir; asıl mühendislik/tasarım emeği zaten prompt şablonunu doğru kurgulamakta harcandı.

Aşağıdaki `llm_ciktilari` sözlüğü, bu manuel süreçte Claude.ai'den alınan gerçek yanıtları içeriyor.

In [62]:
risk_probabilities = rf.predict_proba(X_test)[:, 1]
test_customer_ids = features.loc[X_test.index, 'Customer ID']

risk_summary = X_test.copy()
risk_summary['Customer ID'] = test_customer_ids
risk_summary['risk_probability'] = risk_probabilities
risk_summary['gercek_etiket'] = y_test.values

top_risk = risk_summary.sort_values('risk_probability', ascending=False).head(5)
print(top_risk[['Customer ID', 'risk_probability', 'total_spend', 'past_cancel_rate', 'recency_days', 'total_orders', 'distinct_products']].to_string(index=False))

 Customer ID  risk_probability  total_spend  past_cancel_rate  recency_days  total_orders  distinct_products
       16558          0.860629      5019.86          0.042296             8            19                183
       14911          0.855266     85626.75          0.038988             0           169               1481
       12471          0.851533     10981.36          0.152047             8            36                118
       16161          0.835830      3731.78          0.044510             8            19                205
       13694          0.826761     49347.54          0.030426             1            50                328


top_risk tablosundaki 5 müşterinin her biri için, doldurulmuş tam bir prompt metni üretelim. Bunu kullanma amacım aslında bunu biz yapmayabilirdik. Bir apı key satın alıp bir süreliğine bizim için claude'nin bunu yapmasını isteyebilirdik. Ama hem ücretsiz bir yöntem olduğu için hem de prompt yazma pratiği için bu yolu seçtim. 

Bu kod sayesinde üretilen promptları Claude.ai'ye kopyala-yapıştır yapabiliriz.

Döngüdeki tekrar, "her seferinde yeni bir prompt yazma pratiği" değil, "aynı prompt şablonunu farklı verilerle otomatik doldurma" pratiği 

In [63]:
for i, row in top_risk.iterrows():
    print("="*50)
    print(f"MUSTERI: {row['Customer ID']}")
    print("="*50)
    print(f"""Sen bir e-ticaret operasyon ekibi için çalışan kıdemli bir analistsin.

GOREV: Aşağıdaki müşteri verisini incele ve bu müşterinin neden yüksek iptal/iade riski taşıdığını, operasyon ekibinin anlayabileceği şekilde açıkla.

MUSTERI VERISI:
- Musteri ID: {row['Customer ID']}
- Risk Olasiligi: %{row['risk_probability']*100:.0f}
- Toplam Harcama: £{row['total_spend']:.0f}
- Gecmis Iptal Orani: %{row['past_cancel_rate']*100:.1f}
- Son Alisverisinin Uzerinden Gecen Gun: {row['recency_days']}
- Toplam Siparis Sayisi: {row['total_orders']}
- Farkli Urun Cesidi: {row['distinct_products']}

CIKTI FORMATI:
1. Risk Nedeni (1-2 cumle, hangi ozellikler bu riski isaret ediyor)
2. Onerilen Aksiyon (operasyon ekibinin yapabilecegi somut, tek bir oneri)

KISIT: Toplam yanit 60 kelimeyi gecmesin. Teknik jargon kullanma, operasyon ekibi ML terimlerini bilmiyor.""")
    print()

MUSTERI: 16558.0
Sen bir e-ticaret operasyon ekibi için çalışan kıdemli bir analistsin.

GOREV: Aşağıdaki müşteri verisini incele ve bu müşterinin neden yüksek iptal/iade riski taşıdığını, operasyon ekibinin anlayabileceği şekilde açıkla.

MUSTERI VERISI:
- Musteri ID: 16558.0
- Risk Olasiligi: %86
- Toplam Harcama: £5020
- Gecmis Iptal Orani: %4.2
- Son Alisverisinin Uzerinden Gecen Gun: 8.0
- Toplam Siparis Sayisi: 19.0
- Farkli Urun Cesidi: 183.0

CIKTI FORMATI:
1. Risk Nedeni (1-2 cumle, hangi ozellikler bu riski isaret ediyor)
2. Onerilen Aksiyon (operasyon ekibinin yapabilecegi somut, tek bir oneri)

KISIT: Toplam yanit 60 kelimeyi gecmesin. Teknik jargon kullanma, operasyon ekibi ML terimlerini bilmiyor.

MUSTERI: 14911.0
Sen bir e-ticaret operasyon ekibi için çalışan kıdemli bir analistsin.

GOREV: Aşağıdaki müşteri verisini incele ve bu müşterinin neden yüksek iptal/iade riski taşıdığını, operasyon ekibinin anlayabileceği şekilde açıkla.

MUSTERI VERISI:
- Musteri ID: 14911.0


Claude için Promptlar

MUSTERI: 16558.0
1. Risk Nedeni: Bu müşteri son 8 gündür alışveriş yapmamış olmasına rağmen 19 farklı siparişte 183 farklı ürün denemiş — yüksek ürün çeşitliliği, beğenmediğini iade etme eğilimine işaret ediyor. Toplam harcaması (£5.020) orta düzeyde, güçlü bir sadakat göstergesi değil.

2. Önerilen Aksiyon: Bu müşteriye, son aldığı ürünlerle ilgili bir "doğru beden/kullanım rehberi" içeren kişiselleştirilmiş bir takip e-postası gönderilsin — iade kararını vermeden önce ek bilgiyle desteklenmiş olur.

14911 Numaralı Müşteri İçin Yanıt

1. Risk Nedeni: Bu müşteri şirketin en değerli alıcılarından biri (£85.627 harcama, 169 sipariş) ama aynı zamanda 1.481 farklı ürün denemiş — bu son derece yüksek çeşitlilik, "geniş seçenek deneyip bir kısmını iade etme" alışkanlığına işaret ediyor. Bugün alışveriş yapmış olması, aktif bir döngüde olduğunu gösteriyor.

2. Önerilen Aksiyon: Bu müşteriye özel bir hesap yöneticisi ataması ya da kişiselleştirilmiş ürün önerisi hizmeti sunulsun — hem iade oranını azaltır hem de bu yüksek değerli müşteriyi elde tutmaya katkı sağlar.

12471 Numaralı Müşteri İçin Yanıt

1. Risk Nedeni: Bu müşterinin geçmiş iptal oranı (%15.2), diğer müşterilere kıyasla belirgin şekilde yüksek — yani daha önce de sipariş iptal/iade etme alışkanlığı var. Orta düzey harcama (£10.981) ve 36 siparişlik düzenli bir alışveriş geçmişi, bu davranışın tesadüfi değil, süregelen bir eğilim olduğunu gösteriyor.

2. Önerilen Aksiyon: Bu müşterinin geçmiş iptallerinin hangi ürün kategorilerinde yoğunlaştığı incelensin; eğer belirli bir kategori öne çıkıyorsa (ör. beden/renk uyumsuzluğu), sipariş öncesi o kategoriye özel netleştirici bilgi (ölçü tablosu, detaylı görsel) sunulsun.

16161 Numaralı Müşteri İçin Yanıt

1. Risk Nedeni: Bu müşteri 19 siparişte 205 farklı ürün denemiş — sipariş başına ürün çeşitliliği oldukça yüksek, bu da "aynı üründen birden fazla seçenek sipariş edip bir kısmını iade etme" davranışına işaret ediyor. Nispeten düşük toplam harcama (£3.732), henüz markaya tam bağlanmamış bir profil olabileceğini gösteriyor.

2. Önerilen Aksiyon: Sipariş sırasında "birden fazla beden/renk sipariş edenler için ücretsiz kolay iade" yerine, doğru seçimi ilk seferde yapmasına yardımcı olacak bir beden/uyum önerisi aracı sunulsun.

13694 Numaralı Müşteri İçin Yanıt

1. Risk Nedeni: Bu müşteri yüksek değerli bir alıcı (£49.348 harcama, 50 sipariş) ve dün alışveriş yapmış — aktif bir dönemde. Ancak 328 farklı ürün denemesi, geniş kapsamlı seçim yapıp bir kısmını elden çıkarma eğilimine işaret ediyor; bu, 14911 numaralı müşteriyle benzer bir davranış kalıbı.

2. Önerilen Aksiyon: Bu müşteriye, geçmiş satın alımlarına dayalı daha isabetli/kişiselleştirilmiş ürün önerileri sunularak "deneme amaçlı geniş sipariş verme" ihtiyacı azaltılsın.

Elimizde artık 5 müşteri için risk skoru + doğal dil açıklaması + aksiyon önerisi var. Bunu notebook'ta nasıl belgeleyeceğimize karar verelim.

In [64]:
llm_ciktilari = {
    16558: {
        'risk_nedeni': "Son 8 gündür alışveriş yapmamasına rağmen 19 siparişte 183 farklı ürün denemiş, yüksek ürün çeşitliliği iade eğilimine işaret ediyor. Orta düzey harcaması güçlü bir sadakat göstergesi değil.",
        'aksiyon': "Doğru beden/kullanım rehberi içeren kişiselleştirilmiş takip e-postası gönderilsin."
    },
    14911: {
        'risk_nedeni': "En değerli müşterilerden biri (£85.627, 169 sipariş) ama 1.481 farklı ürün denemesi geniş seçenek deneyip iade etme alışkanlığına işaret ediyor.",
        'aksiyon': "Özel hesap yöneticisi ataması veya kişiselleştirilmiş ürün önerisi hizmeti sunulsun."
    },
    12471: {
        'risk_nedeni': "Geçmiş iptal oranı (%15.2) diğerlerine göre belirgin yüksek, süregelen bir davranış eğilimi gösteriyor.",
        'aksiyon': "Geçmiş iptallerin yoğunlaştığı ürün kategorisi incelenip, o kategoriye özel netleştirici bilgi sunulsun."
    },
    16161: {
        'risk_nedeni': "19 siparişte 205 farklı ürün denemesi yüksek çeşitlilik/deneme davranışına işaret ediyor, düşük toplam harcama marka bağlılığının henüz oluşmadığını gösteriyor.",
        'aksiyon': "Sipariş sırasında beden/uyum önerisi aracı sunularak ilk seferde doğru seçim yapması desteklensin."
    },
    13694: {
        'risk_nedeni': "Yüksek değerli ve aktif bir müşteri (£49.348, dün alışveriş yaptı) ama 328 farklı ürün denemesi geniş kapsamlı deneme davranışına işaret ediyor.",
        'aksiyon': "Geçmiş alımlara dayalı kişiselleştirilmiş öneriler sunularak deneme amaçlı geniş sipariş verme ihtiyacı azaltılsın."
    }
}



In [65]:
llm_df = pd.DataFrame(llm_ciktilari).T.reset_index().rename(columns={'index': 'Customer ID'})
print(llm_df.to_string(index=False))

llm_df.to_csv('llm_risk_ozetleri.csv', index=False)

 Customer ID                                                                                                                                                                                    risk_nedeni                                                                                                             aksiyon
       16558 Son 8 gündür alışveriş yapmamasına rağmen 19 siparişte 183 farklı ürün denemiş, yüksek ürün çeşitliliği iade eğilimine işaret ediyor. Orta düzey harcaması güçlü bir sadakat göstergesi değil.                                 Doğru beden/kullanım rehberi içeren kişiselleştirilmiş takip e-postası gönderilsin.
       14911                                               En değerli müşterilerden biri (£85.627, 169 sipariş) ama 1.481 farklı ürün denemesi geniş seçenek deneyip iade etme alışkanlığına işaret ediyor.                                Özel hesap yöneticisi ataması veya kişiselleştirilmiş ürün önerisi hizmeti sunulsun.
       12471                            

llm_risk_ozetleri.csv dosyası, projenin üç katmanının birleştiği son nokta:

1. ML modeli (Random Forest) → her müşteri için bir risk skoru üretti
2. Prompt sistemi → bu skorları + davranışsal verileri, LLM'in anlayacağı bir formata döktü
3. LLM (Claude) → bu veriyi yorumlayıp, doğal dilde risk nedeni + aksiyon önerisi üretti
4. bu üç adımı birleştirip, sonucu kalıcı bir CSV dosyasına kaydettik

**Threshold Analizi:Eşik Analizi** 

*Modelimiz aslında her müşteri için kesin bir "evet/hayır" değil, bir olasılık üretiyor*

*(predict_proba() ile, ör. "%86 ihtimalle iptal edecek").*

*Bu olasılığı, "riskli" (1) ya da "riskli değil" (0) diye ikiye ayırmak için bir kesme noktasına (threshold) ihtiyacımız var.*

**Varsayılan olarak:** **scikit-learn'ün predict() fonksiyonu**, otomatik olarak **%50'yi eşik alır** — olasılık **%50'nin üzerindeyse 1**, **altındaysa 0 der**.


**Bu %50, rastgele seçilmiş bir sayı — hiçbir iş gerekçesine dayanmıyor, sadece matematiksel bir varsayılan. Threshold Analizi, "gerçekten %50 mi doğru nokta, yoksa başka bir değer bizim iş hedefimize daha mı uygun?" sorusunu sistematik olarak cevaplamaya çalışıyor.**

from sklearn.metrics import precision_score, recall_score, f1_score

# Modelin urettigi olasilik skorlarini al (0/1 degil, yuzde)
risk_probabilities_all = rf.predict_proba(X_test)[:, 1]

# Denenecek farkli threshold degerleri
thresholds = [0.2, 0.3, 0.4, 0.5, 0.6, 0.7]

print(f"{'Threshold':<12}{'Precision':<12}{'Recall':<12}{'F1-Score':<12}")
print("-" * 48)

for t in thresholds:
    ##Farklı eşik değerlerini dene (0.2, 0.3, 0.4, 0.5, 0.6, 0.7 gibi) — her birinde, o eşiğin üzerindeki müşterileri "riskli" say
    y_pred_t = (risk_probabilities_all >= t).astype(int)
    #Her eşikte performansı ölç (precision, recall, F1) — tabloyu oluştur
    precision = precision_score(y_test, y_pred_t)
    recall = recall_score(y_test, y_pred_t)
    f1 = f1_score(y_test, y_pred_t)
    print(f"{t:<12}{precision:<12.2f}{recall:<12.2f}{f1:<12.2f}")


#Sonuçları karşılaştırarak en uygun eşiği seç — bu seçim, sadece matematiksel bir optimizasyon değil, iş ihtiyacına göre verilen bir karar


*%50'de recall sadece %52'ydi — yani riskli müşterilerin neredeyse yarısını kaçırıyorduk. Bu, projenin temel amacına **(riskli müşterileri önceden yakalamak)** aykırı bir sonuçtu. **Threshold analizi** yapmadan, bu sorunu fark etmeyecek ve "model çalışıyor" diye varsayacaktık — oysa gerçekte işe yaramayan bir varsayılan ayarla çalışıyorduk.*

**Somut Katkısı:** 0.4'e Geçince Ne Değişti?
**%50 eşikte:** *Recall %52 (riskli müşterilerin yarısını kaçırıyoruz)*
**%40 eşikte:** *Recall %80 (riskli müşterilerin çoğunu yakalıyoruz)*

Bu, **projenin iş değerini doğrudan artıran somut bir iyileştirme** — **aynı model, aynı veri, sadece karar noktasını iş ihtiyacına göre kalibre ederek, çok daha kullanılabilir bir sistem elde ettik.**

**Seçilen Threshold'u Sisteme Entegre Edelim**

*Şu ana kadar **top_risk** tablosunu oluştururken, aslında threshold'u hiç kullanmıyorduk — sadece olasılıkları büyükten küçüğe sıralayıp ilk 5'i alıyorduk. Şimdi bunu, **gerçekten "riskli" sayılan tüm müşterileri (0.4 eşiğinin üzerindekiler) görecek şekilde güncelleyelim** — bu, gerçek bir üretim sistemine çok daha yakın bir yaklaşım.*

In [68]:
# Secilen threshold: 0.4 (Task 1-2'deki analize dayanarak)
SECILEN_THRESHOLD = 0.4


In [69]:
# Risk olasiliklarini al
risk_probabilities_all = rf.predict_proba(X_test)[:, 1]
test_customer_ids = features.loc[X_test.index, 'Customer ID']

In [70]:
# Tam risk tablosu (butun test seti icin)
full_risk_summary = X_test.copy()
full_risk_summary['Customer ID'] = test_customer_ids
full_risk_summary['risk_probability'] = risk_probabilities_all
full_risk_summary['gercek_etiket'] = y_test.values

In [72]:
# YENI: Threshold'a gore riskli/riskli degil etiketi
import numpy as np
full_risk_summary['risk_durumu'] = np.where(
    full_risk_summary['risk_probability'] >= SECILEN_THRESHOLD, 
    'Riskli', 
    'Riskli Degil'
)

In [73]:
# Kac musteri artik "riskli" olarak isaretleniyor (0.4 esikle)
print("Threshold = 0.4 ile riskli isaretlenen musteri sayisi:")
print(full_risk_summary['risk_durumu'].value_counts())
print()

Threshold = 0.4 ile riskli isaretlenen musteri sayisi:
risk_durumu
Riskli          220
Riskli Degil    158
Name: count, dtype: int64



In [75]:
# Sadece riskli olanlari, risk skoruna gore sirala
riskli_musteriler = full_risk_summary[full_risk_summary['risk_durumu'] == 'Riskli'].sort_values(
    'risk_probability', ascending=False
)

print(f"\nToplam riskli musteri sayisi: {len(riskli_musteriler)}")
print(riskli_musteriler[['Customer ID', 'risk_probability', 'total_spend', 'past_cancel_rate']].head(10).to_string(index=False))


Toplam riskli musteri sayisi: 220
 Customer ID  risk_probability  total_spend  past_cancel_rate
       16558          0.860629      5019.86          0.042296
       14911          0.855266     85626.75          0.038988
       12471          0.851533     10981.36          0.152047
       16161          0.835830      3731.78          0.044510
       13694          0.826761     49347.54          0.030426
       16133          0.823268     11925.28          0.075758
       13069          0.819830      3460.36          0.056410
       17865          0.812809      8202.65          0.046875
       14040          0.801237      2675.58          0.032609
       14189          0.800706      3241.90          0.031579


**Önceki yaklaşım (top_risk):** *Sadece "en riskli 5 müşteri kim" diye bakıyorduk — threshold'un hiçbir etkisi yoktu, sadece sıralama vardı.*

**Yeni yaklaşım:** *Artık gerçek bir karar noktası var — **0.4'ün üzerindeki her müşteri "riskli"** listesine giriyor (5 ile sınırlı değil). Bu, gerçek bir operasyon ekibinin göreceği, iş kuralına dayalı bir liste.*

**220 müşteri, artık "riskli"** olarak işaretleniyor — **eskiden (top_risk ile) sadece 5 müşteriye bakıyorduk**, şimdi 378 kişilik test setinin **%58'i** kadar bir grup riskli görünüyor (378 test müşterisinden 220'si).

**Peki 220 müşteri çok fazla değil mi, operasyon ekibi bunların hepsiyle nasıl ilgilenecek?**

*Bu listenin tamamına aynı yoğunlukta aksiyon uygulanmayacak — risk olasılığına göre önceliklendirilmiş bir liste bu. İlk 10-20 müşteri (en yüksek olasılıklı olanlar) için daha yoğun bir aksiyon (kişisel takip), geri kalanlar için daha hafif bir aksiyon (otomatik bilgilendirme e-postası) uygulanabilir.*

In [76]:
def musteri_risk_ozeti_uret(musteri_verisi, gecmis_geri_bildirim=None):
    """
    musteri_verisi: risk tablosundan bir satir (dict formatinda)
    gecmis_geri_bildirim: eger bu ikinci/ucuncu tur ise, onceki geri bildirim metni
    """
    satirlar = [
        "Sen bir e-ticaret operasyon ekibi icin calisan kidemli bir analistsin.",
        "",
        "GOREV: Asagidaki musteri verisini incele ve bu musterinin neden yuksek iptal/iade riski tasidigini, operasyon ekibinin anlayabilecegi sekilde acikla.",
        "",
        "MUSTERI VERISI:",
        f"- Musteri ID: {musteri_verisi['Customer ID']}",
        f"- Risk Olasiligi: %{musteri_verisi['risk_probability']*100:.0f}",
        f"- Toplam Harcama: £{musteri_verisi['total_spend']:.0f}",
        f"- Gecmis Iptal Orani: %{musteri_verisi['past_cancel_rate']*100:.1f}",
        f"- Toplam Siparis Sayisi: {musteri_verisi['total_orders']}",
        f"- Farkli Urun Cesidi: {musteri_verisi['distinct_products']}",
        "",
        "CIKTI FORMATI:",
        "1. Risk Nedeni (1-2 cumle)",
        "2. Onerilen Aksiyon (somut, tek bir oneri)",
        "",
        "KISIT: Toplam yanit 60 kelimeyi gecmesin. Teknik jargon kullanma.",
    ]
    
    if gecmis_geri_bildirim:
        satirlar.append("")
        satirlar.append(f"ONEMLI: Onceki yanitinla ilgili su geri bildirim verildi, buna gore revize et: {gecmis_geri_bildirim}")
    
    return "\n".join(satirlar)


# TUR 1: Ilk yaniti al
ornek_musteri = riskli_musteriler.iloc[0].to_dict()  # en riskli musteri
ilk_prompt = musteri_risk_ozeti_uret(ornek_musteri)
print("=== TUR 1 - PROMPT ===")
print(ilk_prompt)

=== TUR 1 - PROMPT ===
Sen bir e-ticaret operasyon ekibi icin calisan kidemli bir analistsin.

GOREV: Asagidaki musteri verisini incele ve bu musterinin neden yuksek iptal/iade riski tasidigini, operasyon ekibinin anlayabilecegi sekilde acikla.

MUSTERI VERISI:
- Musteri ID: 16558
- Risk Olasiligi: %86
- Toplam Harcama: £5020
- Gecmis Iptal Orani: %4.2
- Toplam Siparis Sayisi: 19
- Farkli Urun Cesidi: 183

CIKTI FORMATI:
1. Risk Nedeni (1-2 cumle)
2. Onerilen Aksiyon (somut, tek bir oneri)

KISIT: Toplam yanit 60 kelimeyi gecmesin. Teknik jargon kullanma.


**YANIT:** 

**1. Risk Nedeni:** *Bu müşteri 19 siparişte 183 farklı ürün denemiş — yüksek ürün çeşitliliği, birden fazla seçenek sipariş edip beğenmediğini iade etme eğilimine işaret ediyor. Toplam harcaması (£5.020) orta düzeyde, güçlü bir marka bağlılığı henüz oluşmamış olabilir.*

**2. Önerilen Aksiyon:** *Bu müşteriye, geçmiş alışverişlerine dayalı kişiselleştirilmiş bir ürün rehberi (beden/renk uyum önerisi) içeren bir e-posta gönderilsin.*

**Burada prompt'un üzerine LLM'in ürettiği "Risk Nedeni" ve "Önerilen Aksiyon" yapılarını görüyoruz. Önerilen aksiyon net değil. Dolayısyla promptu iyileştirmemiz gerekecek.** 

**YENİ PROMPT**

Sen bir e-ticaret operasyon ekibi icin calisan kidemli bir analistsin.

GOREV: Asagidaki musteri verisini incele ve bu musterinin neden yuksek iptal/iade riski tasidigini, operasyon ekibinin anlayabilecegi sekilde acikla.

MUSTERI VERISI:
- Musteri ID: 16558
- Risk Olasiligi: %86
- Toplam Harcama: £5020
- Gecmis Iptal Orani: %4.2
- Toplam Siparis Sayisi: 19
- Farkli Urun Cesidi: 183

CIKTI FORMATI:
1. Risk Nedeni (1-2 cumle)
2. Onerilen Aksiyon (somut, tek bir oneri)

KISIT: Toplam yanit 60 kelimeyi gecmesin. Teknik jargon kullanma.

ONEMLI: Onceki yanitinla ilgili su geri bildirim verildi, buna gore revize et: Onerilen aksiyon (kisisellestirilmis urun rehberi) cok genel kaldi. Su noktalari netlestirerek daha somut bir oneri ver: (1) Musterinin gecmis kac siparisi/kac ay/kac yillik verisi incelenerek rehber olusturulacak, (2) Musterinin harcama araligi (butce siniri) rehbere dahil edilecek mi, (3) Musterinin sabit/tekrar eden urun kategorileri var mi, bu bilgi kullanilacak mi. Yanit hala 60 kelimeyi gecmesin.

**2. YANIT :**  
**1. Risk Nedeni:** *Bu müşteri 19 siparişte 183 farklı ürün denemiş — yüksek ürün çeşitliliği, birden fazla seçenek sipariş edip beğenmediğini iade etme eğilimine işaret ediyor.*

**2. Önerilen Aksiyon:** *Müşterinin son 6 ay/son 10 siparişindeki harcama aralığı ve en çok tekrar eden 2-3 ürün kategorisi analiz edilerek, bu bütçe ve kategori sınırları içinde kalan, kişiselleştirilmiş 3-4 ürün önerisi içeren bir e-posta gönderilsin.*

## Onay/Geri Bildirim Döngüsü Örneği

LinkedIn'de paylaştığım "sıradaki adımlar" arasında yer alan onay döngüsünü, iki farklı müşteri profili üzerinde test ederek doğruladım.

### Test 1: Müşteri 16558 (Orta Değerli Müşteri)

**Tur 1 - İlk Yanıt:**

Risk Nedeni: Bu müşteri 19 siparişte 183 farklı ürün denemiş — yüksek ürün çeşitliliği, birden fazla seçenek sipariş edip beğenmediğini iade etme eğilimine işaret ediyor. Toplam harcaması (£5.020) orta düzeyde, güçlü bir marka bağlılığı henüz oluşmamış olabilir.

Önerilen Aksiyon: Bu müşteriye, geçmiş alışverişlerine dayalı kişiselleştirilmiş bir ürün rehberi (beden/renk uyum önerisi) içeren bir e-posta gönderilsin.

**Geri Bildirim:** Önerilen aksiyon ("kişiselleştirilmiş ürün rehberi") çok genel kaldı. Rehberin hangi veriyle (kaç sipariş/ay geriye gidilerek), müşterinin bütçe aralığı ve tekrar eden ürün kategorileri dikkate alınarak nasıl oluşturulacağı netleştirilmeli.

**Tur 2 - Revize Edilmiş Yanıt:**

Risk Nedeni: Bu müşteri 19 siparişte 183 farklı ürün denemiş — yüksek ürün çeşitliliği, birden fazla seçenek sipariş edip beğenmediğini iade etme eğilimine işaret ediyor.

Önerilen Aksiyon: Müşterinin son 6 ay/son 10 siparişindeki harcama aralığı ve en çok tekrar eden 2-3 ürün kategorisi analiz edilerek, bu bütçe ve kategori sınırları içinde kalan, kişiselleştirilmiş 3-4 ürün önerisi içeren bir e-posta gönderilsin.

---

### Test 2: Müşteri 14911 (Yüksek Değerli Müşteri)

**Tur 1 - İlk Yanıt:**

Risk Nedeni: Bu müşteri şirketin en değerli alıcılarından biri (£85.627 harcama, 169 sipariş) ama 1.481 farklı ürün denemesi, geniş seçenek deneyip bir kısmını iade etme alışkanlığına işaret ediyor.

Önerilen Aksiyon: Bu müşteriye özel bir hesap yöneticisi ataması veya kişiselleştirilmiş ürün önerisi hizmeti sunulsun.

**Geri Bildirim:** Önerilen aksiyon ("özel hesap yöneticisi/kişiselleştirilmiş öneri") çok genel kaldı. Bu müşteri yüksek değerli olduğu için, hangi somut kriterle (örneğin: son X ayın en çok tekrar eden 3 kategorisi, ortalama sepet büyüklüğü) bir öneri paketi oluşturulacağı ve bu hizmetin ne sıklıkla sunulacağı belirtilmeli.

**Tur 2 - Revize Edilmiş Yanıt:**

Risk Nedeni: Bu müşteri şirketin en değerli alıcılarından biri (£85.627 harcama, 169 sipariş) ama 1.481 farklı ürün denemesi, geniş seçenek deneyip bir kısmını iade etme alışkanlığına işaret ediyor.

Önerilen Aksiyon: Son 3 ayda en çok tekrar eden 3 ürün kategorisi ve ortalama sepet büyüklüğü analiz edilerek aylık kişiselleştirilmiş bir öneri paketi hazırlansın; performansı 3 ay sonra iade oranındaki değişime göre değerlendirilsin.

---

### Karşılaştırma ve Gözlem

| | Müşteri 16558 (orta değerli) | Müşteri 14911 (yüksek değerli) |
|---|---|---|
| Tur 1 sorunu | "kişiselleştirilmiş ürün rehberi" (belirsiz) | "özel hesap yöneticisi/kişiselleştirilmiş öneri" (belirsiz) |
| Tur 2 iyileşmesi | Somut zaman aralığı + bütçe/kategori kriteri | Somut zaman aralığı + kategori kriteri + ölçüm periyodu |

İki farklı müşteri profilinde de aynı tür geri bildirim (aksiyonun belirsizliği), LLM'in çıktısını somut ve uygulanabilir hale getirdi. İkinci testte dikkat çekici bir nokta: LLM, geri bildirimde istemediğimiz halde kendiliğinden bir **ölçüm/değerlendirme mekanizması** da ekledi ("performansı 3 ay sonra iade oranındaki değişime göre değerlendirilsin") — bu, geri bildirim döngüsünün bazen istenenin ötesinde bir kalite artışı da sağlayabileceğini gösteriyor.

**Sonuç:** Bu testler, tek seferlik prompt'ların neden yeterli olmadığını ve bir geri bildirim mekanizmasının, farklı müşteri segmentlerinde de tutarlı şekilde çıktı kalitesini artırdığını doğruluyor.